src/main/java/net/neoforged/neoforge/fluids/FluidUnitSpec.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.fluids;
+
+import com.mojang.datafixers.util.Either;
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.DataResult;
+import com.mojang.serialization.codecs.RecordCodecBuilder;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import net.minecraft.network.chat.Component;
+import net.minecraft.util.Mth;
+import org.jetbrains.annotations.Nullable;
+
+public class FluidUnitSpec {
+    public static final Codec<FluidUnitSpec> CODEC = RecordCodecBuilder.create(instance -> instance.group(
+            Codec.unboundedMap(FluidUnit.CODEC, UnitInfo.CODEC).validate(FluidUnitSpec::checkFixedUnits).fieldOf("units").forGetter(s -> s.map),
+            FluidUnit.CODEC.optionalFieldOf("default_unit", FluidUnits.BUCKET).forGetter(FluidUnitSpec::defaultUnit),
+            Codec.INT.optionalFieldOf("smallest_decimals", -1).forGetter(s -> s.smallestUnitDecimals)).apply(instance, FluidUnitSpec::new));
+
+    private final Map<FluidUnit, UnitInfo> map;
+
+    private final int smallestUnitDecimals;
+    private final FluidUnit defaultUnit;
+    private final List<FluidUnit> listForTooltip;
+
+    public static final FluidUnitSpec DEFAULT = new FluidUnitSpec(
```

Comment:
What does this default value represent? It's not immediately clear from looking at it.

Reply:
Looks to me like its the fallback if units are not defined for a fluid, which just displays in buckets with 3 decimal points below buckets. So you get values like `1.345 buckets`.